src/xwayland.c

```diff
@@ -880,3 +880,74 @@ xwayland_server_finish(struct server *server)
 	server->xwayland = NULL;
 	wlr_xwayland_destroy(xwayland);
 }
+
+void
+xwayland_update_workarea(struct server *server)
+{
+	struct wlr_box lb;
+	wlr_output_layout_get_box(server->output_layout, NULL, &lb);
+
+	/* Compute outer edges of layout (excluding negative regions) */
+	int layout_left = MAX(0, lb.x);
+	int layout_right = MAX(0, lb.x + lb.width);
+	int layout_top = MAX(0, lb.y);
+	int layout_bottom = MAX(0, lb.y + lb.height);
+
+	/* Workarea is initially the entire layout */
+	int workarea_left = layout_left;
+	int workarea_right = layout_right;
+	int workarea_top = layout_top;
+	int workarea_bottom = layout_bottom;
+
+	struct output *output;
+	wl_list_for_each(output, &server->outputs, link) {
+		if (!output_is_usable(output)) {
+			continue;
+		}
+
+		struct wlr_box ob;
+		wlr_output_layout_get_box(server->output_layout,
+			output->wlr_output, &ob);
+
+		/* Compute edges of output */
+		int output_left = ob.x;
+		int output_right = ob.x + ob.width;
+		int output_top = ob.y;
+		int output_bottom = ob.y + ob.height;
+
+		/* Compute edges of usable area */
+		int usable_left = output_left + output->usable_area.x;
+		int usable_right = usable_left + output->usable_area.width;
+		int usable_top = output_top + output->usable_area.y;
+		int usable_bottom = usable_top + output->usable_area.height;
+
+		/*
+		 * Only adjust workarea edges for output edges that are
+		 * aligned with outer edges of layout
+		 */
+		if (output_left == layout_left) {
+			workarea_left = MAX(workarea_left, usable_left);
+		}
+		if (output_right == layout_right) {
+			workarea_right = MIN(workarea_right, usable_right);
+		}
+		if (output_top == layout_top) {
+			workarea_top = MAX(workarea_top, usable_top);
+		}
+		if (output_bottom == layout_bottom) {
+			workarea_bottom = MIN(workarea_bottom, usable_bottom);
+		}
+	}
+
+	/*
+	 * Set _NET_WORKAREA property. We don't report virtual desktops
+	 * to XWayland, so we set only one workarea.
+	 */
```

Comment:
This is something that I'd actually like to change at some point.

Reply:
That would be cool. I'm not certain exactly how that would affect this code -- maybe it would just send multiple copies of the same wlr_box, maybe no change at all. Usable area is the same for all workspaces in labwc, I think?

(Openbox may also make a distinction between "virtual desktops" and "workspaces". I recall this vaguely from a discussion several years back, so I would have to refresh myself on the details.)

Reply:
> (Openbox may also make a distinction between "virtual desktops" and "workspaces". I recall this vaguely from a discussion several years back, so I would have to refresh myself on the details.)

Maybe there is indeed some difference in the X11 terminology, I've always considered them to be synonyms but that might actually be wrong and one of them is what we call "output"?

> Usable area is the same for all workspaces in labwc, I think?

Yes, a workspace spans all outputs.

However, as layershell exclusive zones are per output (same as X11 ones I think? e.g. a panel may only be on the primary monitor) this calculation does't seem always correct. Couldn't we just use `output_usable_area_in_layout_coords()` for each output and throw the returned wlr_box directly into the X11 function?

Reply:
Looking through the EMWH spec, I think "virtual desktops" does mean essentially "workspaces". I don't think putting per-output geometries in _NET_WORKAREA is correct though. It seems EWMH doesn't really consider multiple physical outputs.